---
title: Spontaneous riboflavin-overproducing Limosilactobacillus reuteri for biofortification
  of fermented foods
author: "Sarah Ahannach"
date: "April 14, 2022"
output: html_document
---

This R markdown document describes the data analysis for the paper "Spontaneous riboflavin-overproducing Limosilactobacillus reuteri for biofortification of fermented foods" by Irina Spacaova, Sarah Ahannach, Annelies Breynaert, Isabel Erreygers, Stijn Wittouck, Peter A. Bron, Wannes Van Beeck, Tom Eilers, Abbas Alloul, Na√Øm Blansaer, Siegfried E. Vlaeminck, Nina Hermans, Sarah Lebeer. All data visualizations and summary tables of the microbiome sequencing data shown in the paper are generated by this document. Microbiome data is available in the European Nucleotide Archive (ENA) under accession number PRJEB52196. You can read this document to get a more detailed overview of the data analysis, or you can run the code yourself and play around with it if you want to explore the data on your own. The easiest way to run the code is to use [RStudio](https://www.rstudio.com/products/rstudio/#Desktop). 

--------
Content
--------
- Set-up
- Part 1: Quality Control
- Part 2: 16S rRNA gene amplicon sequencing analysis

------------------------------------------------------------------------------------------------------------------------
Set-up
------------------------------------------------------------------------------------------------------------------------

You can install these packages by uncommenting and running the following commands:
```{r}
#install.packages("tidyverse")
#devtools::install_github("SWittouck/tidyamplicons")
#install.packages("rmarkdown")
#install.packages("devtools")
#install.packages("ggpubr")
#install.packages("vegan")
#install.packages("ggrepel")
```

Settings for rendering of this markdown file:
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Load the required R packages: 
```{r}
library(tidyverse)
library(tidyamplicons)
library(readr)
library(ggpubr)
library(vegan)
library(ggrepel)
```

------------------------------------------------------------------------------------------------------------------------
PART 1
------------------------------------------------------------------------------------------------------------------------

## Data preparation

We start by loading the data. All data is bundled in a "tidyamplicons" object: it contains the read counts of all taxa (ASVs) in all samples, the metadata of the samples (sampling location, swab, method, timepoint,...) and metadata of the taxa (taxonomic classification and interpretable taxon names). 
```{r}
load("../data/ambv339_gidm.rda")
```

Next we calculate the library concentration by making a new column: lib_size/pooled volume = concentration. But don't forget to first add lib_size
```{r}
run <- add_lib_size(run)
run <- mutate_samples (run, concentration = lib_size / pooled_volume)
#View(run$samples)
```


## ASV filtering

### Too long reads

General exploration
```{r}
get_numbers(run)
```

Plot length of ASVs:
```{r,fig.height= 3}
run$taxa <- run$taxa %>%
  mutate(length = str_length(taxon))
run$taxa %>%
  count(length, name ="n_asvs = n") %>%
  ggplot(aes(x =length, y = n_asvs)) +
  geom_bar(stat = "identity") +
  scale_y_log10()

run %>%
  add_lib_size() %>%
  samples()%>%
  ggplot(aes(fct_reorder(sample,lib_size),lib_size)) +
  geom_col()

#if you want descending
run %>%
  add_lib_size() %>%
  samples()%>%
  ggplot(aes(fct_reorder(sample,lib_size, .desc = T),lib_size)) +
  geom_col()
```

Filter out ASVs that are too long:
```{r}
run$taxa <- run$taxa %>%
  filter_taxa(length < 300) %>%
  update_lib_sizes(step = "removing_too_long")

get_numbers(run)
```


### Non-bacterial reads

```{r, message=TRUE, fig.height=3}
run <- add_lib_size(run)
run <- mutate_samples(run, libmcb = lib_size) # libmcb is now the original amount of all reads, including non-bacterial

run$taxa <- run$taxa %>%
  filter_taxa(kingdom == "Bacteria") %>%
  filter_taxa(family !="Mitochondria" | is.na(family)) %>%
  filter_taxa(class != "Chloroplast" | is.na(class))
#add | is.na(...) --> because you don't want to filter our the NA's
run <- add_lib_size(run)

run <- mutate_samples(run, nonbacterial=(libmcb-lib_size)/libmcb)
```

```{r}
run <- update_lib_sizes(run, step = "removing_non_bacterial")
get_numbers(run)
```

Plot non-bacterial per sample
```{r,fig.height=5}
ggplot(run$samples, aes(x = sample_id, y=nonbacterial)) +
  geom_jitter(size = 2) +
  geom_boxplot(alpha = 0.1) +
  theme_bw()
```

Calculate non-bacterial in samples
```{r}
mean(run$samples$nonbacterial, na.rm = T)
```


## Quality control of samples

```{r,fig.height=3}
ggplot(run$lib_sizes, aes(x = lib_size)) +
  scale_x_log10() +
  geom_density() +
  geom_rug()
```

```{r}
ggplot(run$lib_sizes, aes(x = lib_size,  y = sample_id)) +
    geom_jitter(alpha = 0.2)
```

```{r}
ggplot(run$samples, aes(x = lib_size,  y = sample_id)) +
  geom_jitter(alpha = 0.2)
```

```{r}
ggplot(run$samples, aes(x = sample_id, y = lib_size)) +
  geom_jitter(size = 2) +
  geom_boxplot (alpha = 0.1) +
  scale_y_log10() +
  theme_bw()
```

```{r}
run %>%
  add_lib_size() %>%
  samples() %>%
  ggplot(aes(x = sample_id, y = lib_size)) +
  geom_bar(stat = "identity") +
  scale_y_log10() +
  theme_bw()
```

```{r}
run$lib_sizes %>%
  group_by(step) %>%
  summarize(n_reads = sum(lib_size))
```


------------------------------------------------------------------------------------------------------------------------
PART 2
------------------------------------------------------------------------------------------------------------------------

## Separate samplenames into columns

```{r}
run$samples <- run$samples %>%
  separate(description, into = c('owner', 'condition', 'timepoint', 'experiment')) %>%
  arrange(experiment)
#View(run$samples)
```

change content of column names

```{r}
run <- run %>%
  mutate_samples(experiment = recode(experiment,
        "1" = "Saline",
        "2" = "Coconut")) %>%
  mutate_samples(timepoint = recode(timepoint,
        "0" = "0 hours",
        "2" = "2 hours",
        "4" = "4 hours",
        "6" = "6 hours",
        "24" = "24 hours",
        "48" = "48 hours",
        "72" = "72 hours")) %>%
  mutate_samples(condition = recode(condition,
        "A" = "AMBV339+feces (1)",
        "B" = "AMBV339+feces (2)",
        "F" = "Feces_control")) 
#View(run$samples)
```


## Explore contaminants indicated by blanks

```{r,fig.height=3}
run_blanks <- 
  filter_samples(run, condition %in% c("NC"))
get_bar_plot(run_blanks, x = condition)
#View(run_blanks$samples)
```

filter out controls from run

```{r}
run <- run %>%
  filter_samples(condition != "NC")
#View(run$samples)
```


## Calculate relative abundance Limosilactobacillus reuteri AMBV339

```{r}
run %>%
  aggregate_taxa(rank = "genus") %>%
  add_mean_rel_abundances(condition = "condition") %>%
  taxa() %>%
  filter(genus == "Lactobacillus")
```

- we are assuming here that L. reuteri AMBV339 is the only strain of the old genus Lactobacillus present in the data


## Filter out AMBV339

```{r}
run <- run %>%
  filter_taxa(taxon_id !="t2")
#View(run$taxa)
```

Based on taxon table: Lactobacillus antri/frumenti/oris/panis/reuteri
TACGTAGGTGGCAAGCGTTATCCGGATTTATTGGGCGTAAAGCGAGCGCAGGCGGTTGCTTAGGTCTGATGTGAAAGCCTTCGGCTTAACCGAAGAAGTGCATCGGAAACCGGGCGACTTGAGTGCAGAAGAGGACAGTGGAACTCCATGTGTAGCGGTGGAATGCGTAGATATATGGAAGAACACCAGTGGCGAAGGCGGCTGTCTGGTCTGCAACTGACGCTGAGGCTCGAAAGCATGGGTAGCGAACAGG
- taxon id: t2


## Results section: L. reuteri AMBV339 does not perturb the healthy fecal microbiome and inhibits enteric pathogens

*Figure 6A*

```{r,fig.height=5, fig.width=12}
run %>%
  add_taxon_name()%>%
    mutate_samples(condition=factor(condition,levels=c("AMBV339+feces (1)", "AMBV339+feces (2)", "Feces_control")))%>%
  mutate_samples(timepoint=factor(timepoint,levels=c("0 hours", "2 hours", "4 hours", "6 hours", "24 hours", "48 hours", "72 hours")))%>%
  aggregate_taxa(rank = "genus") %>%
get_bar_plot(x = condition) +
  facet_wrap(~timepoint,scales="free",ncol=1) +
  facet_grid(rows = NULL, cols = vars(timepoint), scales="free", space = "free", facets = vars(experiment)) +
  scale_color_brewer(palette = "Paired")+
  theme_bw()+
  theme(legend.position = "right", legend.text = element_text(size = 14), legend.title = element_text(size = 14),axis.title.x = element_blank(), axis.title=element_text(size=14), axis.text=element_text(size=10),strip.text=element_text(size=14))+
  rotate_x_text(angle = 90) 
```

*Supplementary Figure S3A*

```{r,fig.height=5, fig.width=12}
timelvls <- c("0 hours", "2 hours", "4 hours", "6 hours", "24 hours", "48 hours", "72 hours")
conditionlvls <- c("AMBV339+feces (1)", "AMBV339+feces (2)", "Feces_control")

run %>%
  add_alphas() %>%
  samples() %>%
  mutate(condition = factor(condition, levels = conditionlvls)) %>%
  mutate(timepoint = factor(timepoint, levels = timelvls)) %>%
  ggplot(aes(x = condition, y = inverse_simpson, col = condition, group = condition, size = 0.1)) +
  geom_point(size = 2, alpha = 0.8, position = position_dodge(width = 1)) +
  stat_summary(geom = "point", fun = "mean", size = 10, shape = "-", position = position_dodge(width = 1)) +
  facet_wrap(~timepoint,scales="fixed",ncol=7) +
  theme_bw() +
  theme(legend.position = "right", legend.text = element_text(size = 14)) +
  theme(axis.title = element_text(size = 14), axis.text = element_text(size = 10)) +
  labs(y = "inverse simpson") +
  theme(axis.title.x = element_blank()) +
  scale_fill_brewer(palette = "Paired") +
    theme_bw()+
  theme(legend.position = "none", legend.text = element_text(size = 14), legend.title = element_text(size = 14),axis.title.x = element_blank(), axis.title=element_text(size=14), axis.text=element_text(size=10),strip.text=element_text(size=14))+
  rotate_x_text(angle = 90)
```
  
*Supplementary Figure S3B*

```{r,fig.height=5, fig.width=12}
timelvls <- c("0 hours", "2 hours", "4 hours", "6 hours", "24 hours", "48 hours", "72 hours")
conditionlvls <- c("AMBV339+feces (1)", "AMBV339+feces (2)", "Feces_control")

run %>%
  betas()%>%
  filter(condition_1 == condition_2)%>%
  filter(experiment_1 == experiment_2)%>%
  filter(timepoint_1 != timepoint_2)%>%
  filter(timepoint_1 =="0 hours")%>%
  mutate(condition_1=factor(condition_1,levels= conditionlvls))%>%
  mutate(timepoint_2=factor(timepoint_2,levels= timelvls))%>%
  ggplot(aes(x=condition_2,y=beta,col = condition_1, group = condition_1, size = 0.1))+
  geom_point(size = 2, alpha = 0.8, position = position_dodge(width = 1)) +
  stat_summary(geom = "point", fun = "mean", size = 10, shape = "-", position = position_dodge(width = 1)) +
  facet_wrap(~timepoint_2,scales="fixed",ncol=7) +
    theme(axis.title.x = element_blank()) +
  scale_fill_brewer(palette = "Paired") +
  theme_bw()+
  theme(legend.position = "none", legend.text = element_text(size = 14), legend.title = element_text(size = 14),axis.title.x = element_blank(), axis.title=element_text(size=14), axis.text=element_text(size=10),strip.text=element_text(size=14))+
  rotate_x_text(angle = 90)+
  labs(y = "beta diversity")
```

*Supplementary Figure S3C*

```{r,fig.height=7, fig.width=10}
run %>%
    aggregate_taxa(rank = "genus") %>%
    add_rel_abundance() %>%
    get_abundances_extended() %>%
    filter(genus %in% c("Bifidobacterium","Blautia","Clostridium","Faecalibacterium","Romboutsia","Bacteroides"))%>%
    mutate(timepoint=factor(timepoint,levels=c("0 hours", "2 hours", "4 hours", "6 hours", "24 hours", "48 hours", "72 hours")))%>%
    ggplot(aes (x = timepoint , y = rel_abundance, group = timepoint, col = genus)) +
    geom_point(size = 2, alpha = 0.8, position = position_dodge(width = 1)) +
    stat_summary(geom = "point", fun = "mean", size = 10, shape = "-", position = position_dodge(width = 1)) +
    facet_wrap(~genus, scales = "fixed", ncol = 6) +
    facet_grid(rows = NULL, cols = vars(genus), scales="fixed", space = "free", facets = vars(condition))+
    scale_color_brewer(palette = "Paired")+
    ylab ("Relative abundance")+
    theme_bw() +
    theme(strip.text = element_text(size = 12, face = "italic"),
          legend.position = "none",
          axis.title.x = element_text(size=14),
          axis.title.y = element_text(size=14),
          axis.text = element_text(size = 8))+
    rotate_x_text(angle = 90)
    scale_x_discrete(guide = guide_axis(n.dodge = 2))
    scale_y_log10()
```

_____________________________________________________________________________________________________________________

The end
_____________________________________________________________________________________________________________________








